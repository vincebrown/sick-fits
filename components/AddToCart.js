import React from 'react'
import PropTypes from 'prop-types'

import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'

import { CURRENT_USER_QUERY } from './User'

const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`

class AddToCart extends React.Component {
  static propTypes = {
    id: PropTypes.string,
  }

  render() {
    const { id } = this.props
    return (
      <Mutation
        mutation={ADD_TO_CART_MUTATION}
        variables={{
          id,
        }}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(addToCart, { loading }) => (
          <button disabled={loading} type="button" onClick={addToCart}>
            Add{loading && 'ing'} to Cart
          </button>
        )}
      </Mutation>
    )
  }
}

export default AddToCart
export { ADD_TO_CART_MUTATION }
