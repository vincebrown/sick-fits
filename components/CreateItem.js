// Packages
import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Router from 'next/router'

// Components
import ErrorMessage from './ErrorMessage'

// Styled Components
import Form from './styles/Form'

// Libs
import formatMoney from '../lib/formatMoney'

// Query
const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    price: 0,
    image: '',
    largeImage: '',
  }

  _handleChange = e => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({ [name]: val })
  }

  _uploadFile = async e => {
    const { files } = e.target
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', 'sickfits')

    const res = await fetch(
      'https://api.cloudinary.com/v1_1/vincebrown/image/upload',
      {
        method: 'POST',
        body: data,
      }
    )

    const file = await res.json()

    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    })
  }

  render() {
    const { title, price, description, image, largeImage } = this.state

    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            data-test="form"
            onSubmit={async e => {
              // Stop from submitting
              e.preventDefault()
              // Call the mutation
              const res = await createItem()
              // Change them to the single item page
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              })
            }}
          >
            <ErrorMessage error={error} />

            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an Image"
                  required
                  onChange={this._uploadFile}
                />
                {image && <img src={image} alt="Upload Preview" />}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={title}
                  onChange={this._handleChange}
                />
              </label>

              <label htmlFor="title">
                Price {price ? formatMoney(price) : ''}
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={price}
                  onChange={this._handleChange}
                />
              </label>

              <label htmlFor="title">
                Description
                <textarea
                  type="text"
                  id="description"
                  name="description"
                  placeholder="Description"
                  required
                  value={description}
                  onChange={this._handleChange}
                />
              </label>
              <button type="submit">Create Item</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
