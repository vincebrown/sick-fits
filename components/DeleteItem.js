import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'

// eslint-disable-next-line import/no-cycle
import { ALL_ITEMS_QUERY } from './Items'

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`

class DeleteItem extends Component {
  static propTypes = {
    children: PropTypes.node,
    id: PropTypes.string.isRequired,
  }

  _update = (cache, payload) => {
    // manually update cache on client
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY })
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    )
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data })
  }

  render() {
    const { children, id } = this.props

    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id }}
        update={this._update}
      >
        {(deleteItem, { error }) => (
          <button
            onClick={() => {
              if (window.confirm('Are you sure you want to delete?')) {
                deleteItem().catch(err => {
                  alert(err.message)
                })
              }
            }}
            type="button"
          >
            {children}
          </button>
        )}
      </Mutation>
    )
  }
}

export default DeleteItem
