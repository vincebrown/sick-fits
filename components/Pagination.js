import React from 'react'
import propTypes from 'prop-types'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'
import Head from 'next/head'
import Link from 'next/link'
import PaginationStyles from './styles/PaginationStyles'
import { perPage } from '../config'

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`
const Pagination = props => (
  <Query query={PAGINATION_QUERY}>
    {({ error, data, loading }) => {
      const { itemsConnection } = data
      const { count } = itemsConnection.aggregate
      const pages = Math.ceil(count / perPage)

      if (loading) return <p>Loading.....</p>

      return (
        <PaginationStyles>
          <Head>
            <title>
              Sick Fits! Page {props.page} of {pages}
            </title>
          </Head>
          <Link
            prefetch
            href={{ pathname: 'items', query: { page: props.page - 1 } }}
          >
            <a className="prev" aria-disabled={props.page <= 1}>
              Prev
            </a>
          </Link>
          <p>
            Page {props.page} of {pages}
          </p>
          <p>{count} Items Total</p>
          <Link
            prefetch
            href={{ pathname: 'items', query: { page: props.page + 1 } }}
          >
            <a className="next" aria-disabled={props.page >= pages}>
              Next
            </a>
          </Link>
        </PaginationStyles>
      )
    }}
  </Query>
)

Pagination.propTypes = {
  page: propTypes.number,
}

export default Pagination
