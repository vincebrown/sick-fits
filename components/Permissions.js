import React, { Component } from 'react'
import { Query, Mutation } from 'react-apollo'
import PropTypes from 'prop-types'
import gql from 'graphql-tag'
import Error from './ErrorMessage'
import Table from './styles/Table'
import SickButton from './styles/SickButton'

const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
]

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    users {
      id
      name
      email
      permissions
    }
  }
`

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation UPDATE_PERMISSIONS_MUTATION(
    $permissions: [Permission]
    $userId: ID!
  ) {
    updatePermissions(permissions: $permissions, userId: $userId) {
      id
      name
      email
      permissions
    }
  }
`

const Permissions = () => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) => {
      if (error) return <Error error={error} />
      if (loading) return <p>Loading...</p>

      return (
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermissions.map(permission => (
                  <th key={`${permission}-head`}>{permission}</th>
                ))}
                <th>ðŸ‘‡</th>
              </tr>
            </thead>
            <tbody>
              {data.users.map(user => (
                <UserPermissions key={`${user.id}-permission`} user={user} />
              ))}
            </tbody>
          </Table>
        </div>
      )
    }}
  </Query>
)

class UserPermissions extends Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      permissions: PropTypes.array,
    }).isRequired,
  }

  constructor(props) {
    super(props)

    const { user } = props

    this.state = {
      permissions: user.permissions,
    }
  }

  handlePermissionChange = e => {
    const { permissions } = this.state
    const checkbox = e.target
    let updatedPermissions = [...permissions]
    if (checkbox.checked) {
      updatedPermissions.push(checkbox.value)
    } else {
      updatedPermissions = updatedPermissions.filter(
        permission => permission !== checkbox.value
      )
    }
    this.setState({ permissions: updatedPermissions })
  }

  render() {
    const { user } = this.props
    const { permissions } = this.state
    return (
      <Mutation
        mutation={UPDATE_PERMISSIONS_MUTATION}
        variables={{
          permissions,
          userId: user.id,
        }}
      >
        {(updatePermissions, { loading, error }) => (
          <>
            {error && <Error error={error} />}
            <tr>
              <td>{user.name}</td>
              <td>{user.email}</td>
              {possiblePermissions.map(permission => (
                <td key={`${permission}-key`}>
                  <label htmlFor={`${user.id}-permission${permission}`}>
                    <input
                      type="checkbox"
                      id={`${user.id}-permission${permission}`}
                      checked={permissions.includes(permission)}
                      value={permission}
                      onChange={this.handlePermissionChange}
                    />
                  </label>
                </td>
              ))}
              <td>
                <SickButton
                  type="button"
                  disabled={loading}
                  onClick={updatePermissions}
                >
                  Updat{loading ? 'ing' : 'e'}
                </SickButton>
              </td>
            </tr>
          </>
        )}
      </Mutation>
    )
  }
}

export default Permissions
