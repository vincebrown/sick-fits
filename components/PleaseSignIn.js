import { Query } from 'react-apollo'
import PropTypes from 'prop-types'

import { CURRENT_USER_QUERY } from './User'
import Signin from './Signin'

const PleaseSignIn = props => (
  <Query query={CURRENT_USER_QUERY}>
    {({ data, loading }) => {
      if (loading) return <p>Loading..</p>

      if (!data.me) {
        return (
          <div>
            <h1>Please Sign in before continuing</h1>
            <Signin />
          </div>
        )
      }

      return props.children
    }}
  </Query>
)

PleaseSignIn.propTypes = {
  children: PropTypes.node.isRequired,
}
export default PleaseSignIn
