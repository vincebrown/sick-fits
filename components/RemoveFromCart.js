import React from 'react'
import PropTypes from 'prop-types'
import { Mutation } from 'react-apollo'
import styled from 'styled-components'
import gql from 'graphql-tag'

import { CURRENT_USER_QUERY } from './User'

const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: none;
  &:hover {
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`

class RemoveFromCart extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  }

  // This gets called as soon as we get a response
  // Back from the server that mutation happened.
  update = (cache, payload) => {
    // read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY })
    const itemId = payload.data.removeFromCart.id
    // remove item from the cart.
    data.me.cart = data.me.cart.filter(item => item.id !== itemId)
    // write it back to cache.
    cache.writeQuery({ query: CURRENT_USER_QUERY, data })
  }

  render() {
    const { id } = this.props
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        update={this.update}
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id,
          },
        }}
        variables={{
          id,
        }}
      >
        {(removeFromCart, { error, loading }) => (
          <BigButton
            disabled={loading}
            role="button"
            title="Delete Item"
            onClick={() => {
              removeFromCart().catch(err => alert(err.message))
            }}
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    )
  }
}

export default RemoveFromCart
export { REMOVE_FROM_CART_MUTATION }
