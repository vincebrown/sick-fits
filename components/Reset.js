import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'

import Form from './styles/Form'
import Error from './ErrorMessage'

import { CURRENT_USER_QUERY } from './User'

const RESET_MUTATION = gql`
  mutation RESET_MUTATION(
    $resetToken: String!
    $newPassword: String!
    $confirmPassword: String!
  ) {
    resetPassword(
      resetToken: $resetToken
      newPassword: $newPassword
      confirmPassword: $confirmPassword
    ) {
      id
      name
      email
    }
  }
`

export default class Reset extends Component {
  static propTypes = {
    resetToken: PropTypes.string.isRequired,
  }

  state = {
    newPassword: '',
    confirmPassword: '',
  }

  saveToState = e => {
    this.setState({ [e.target.name]: e.target.value })
  }

  render() {
    const { newPassword, confirmPassword } = this.state
    const { resetToken } = this.props
    return (
      <Mutation
        mutation={RESET_MUTATION}
        variables={{
          resetToken,
          newPassword,
          confirmPassword,
        }}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(resetPassword, { error, loading, called }) => (
          <Form
            method="post"
            onSubmit={async e => {
              e.preventDefault()
              await resetPassword(newPassword, confirmPassword)
              this.setState({ newPassword: '', confirmPassword: '' })
            }}
          >
            <fieldset disabled={loading} aria-busy={loading}>
              <h2>Reset Password</h2>
              <Error error={error} />
              {!error && !loading && called && (
                <p>Success! Your Password has been reset</p>
              )}
              <label htmlFor="newPassword">
                New Password
                <input
                  type="password"
                  name="newPassword"
                  placeholder="Enter a new Password"
                  value={newPassword}
                  onChange={this.saveToState}
                />
              </label>
              <label htmlFor="confirmPassword">
                Confirm Password
                <input
                  type="password"
                  name="confirmPassword"
                  placeholder="Confirm your new password"
                  value={confirmPassword}
                  onChange={this.saveToState}
                />
              </label>

              <button type="submit">Reset Password</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}
