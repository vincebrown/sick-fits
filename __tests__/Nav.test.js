// Packages
import { mount } from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import { MockedProvider } from 'react-apollo/test-utils'

// Utils
import { fakeUser } from '../lib/testUtils'

// Components
import Nav from '../components/Nav'
import { CURRENT_USER_QUERY } from '../components/User'

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
]

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
]

describe('<Nav/>', () => {
  it('renders a minimal nav when signed out', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    const nav = wrapper.find('ul[data-test="nav"]')
    expect(toJSON(nav)).toMatchSnapshot()
  })

  it('renders full nav when signed in', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    const nav = wrapper.find('ul[data-test="nav"]')
    expect(nav.children().length).toBe(6)
    expect(nav.text()).toContain('Sign Out')
  })
})
