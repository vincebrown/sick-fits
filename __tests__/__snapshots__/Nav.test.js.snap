// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
<ul
  className="NavStyles-sc-1ry6i01-0 kSFkjr"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign In
    </a>
  </Link>
</ul>
`;
